### ISSUE TITLE
Remove jQuery CDN options

### ISSUE THREAD

**Issue by danieliser** [roots/soil#155]
This feature is a bad call in any situation, short of writing every function yourself this is a surefire way to break something somewhere.

https://ttmm.io/tech/dont-dequeue-wordpress-jquery/

Too many themes already do this against guidelines, why would a respected theme framework be promoting this idea further? If someone wants to box themselves in, I'm sure they can find it easy enough, but for you to have it on by default in the sage theme is a terrible idea. Don't make it easier for them to break stuff, this just promotes releasing of this type of thing into the wild for mass consumption.

As a plugin developer I see this and it makes me cringe based on the literally hundreds of support tickets that could have been avoided just by theme authors not doing this. Would have 0 effect on their themes capabilities, but they do it because they see it done other places.

Please don't perpetuate bad habits inside a great framework that people follow.

**Comment by retlehs** [comment]
> This feature is a bad call in any situation, short of writing every function yourself this is a surefire way to break something somewhere.

in all the **years** that we have had this feature we have encountered no issues. soil enqueues the same version of jquery that wordpress uses.

> As a plugin developer I see this and it makes me cringe based on the literally hundreds of support tickets that could have been avoided just by theme authors not doing this

please provide specifics on a scenario where soil's jquery cdn has broke something. in the meantime, stop spreading FUD.

**Comment by danieliser** [comment]
@retlehs - Just because you don't have or see an issue directly doesn't mean it's not rampant somewhere else.

And I never meant to insenuate Soil and Roots were the theme causing my headaches directly. But this is a growing issue in WP themes despite so many people advising agaist it. People you would never just close the issue on.



You may at this exact moment have the same version but what happens when WP core changes versions.

In any case by keeping this you are teaching that it's not only ok but good practice. Undoing years of work by others to undo these support nightmares.

PS you haven't heard any issues because the user never places the blame where it actually goes. Over 60% of my plugins 200+ tickets a week deal with other theme and plugin authors doing stuff like this. User thinks it's great until it breaks something at which point they generally don't even consider that their theme is responsible.

So what I am asking is for you to seriously consider the implications of a popular theme framework enabling this by default. You are literally telling authors this is not only a good idea it should be in everything.

So for all the thousands of plugin authors that have ever dealt with even a single issue caused by loading a custom jQuery I want to thank you for seeing that our time too is valuable.

**Comment by retlehs** [comment]
so... you literally have no example of soil's jquery cdn implementation causing any sort of an issue, yet you demand that we remove the feature?

**Comment by kalenjohnson** [comment]
> You may at this exact moment have the same version but what happens when WP core changes versions.

https://github.com/roots/soil/blob/master/modules/jquery-cdn.php#L12

Shouldn't be an issue :wink:

**Comment by danieliser** [comment]
Just so you know, I was not trying to attack you or the products. Love what you guys have done as so many others do. But you said it all when you closed the issue without even a dialougue. As I said. It's considered bad practice in general. So is that something you want to be pushing out. It's pure and simple not in the spirit of WP nor was your initial reaction.

I simply want a valid discussion on the matter because your framework is becoming so popular others will inevitably look at your sites and themes and decide that's the way they are gonna do it. They may not be as thorough as you, they may load any version that comes up with the first Google link for CDN jQuery. 

Point is mute. If it's considered bad practice in general. Why promote it.

Read the first paragraph of the actual guidelines.

https://developer.wordpress.org/themes/basics/including-css-javascript/

It says remember your not the only active plugin/theme. Be considerate. And that is specifically talking about JS and CSS assets.

And if that's not good enough here is another specifically saying don't do it in big red box.

https://developer.wordpress.org/themes/advanced-topics/javascript-best-practices/

Just because you can doesn't mean you should and like I said, I'm not concerned about what you see directly. I am concerned with what you are promoting in general.

**Comment by retlehs** [comment]
considered bad practice by who, wordpress.org? **wordpress is filled with bad practices** — take wpdev advice with a grain of salt. 

> They may not be as thorough as you, they may load any version that comes up with the first Google link for CDN jQuery.

other people doing it wrong has nothing to do with us...

btw you keep mentioning the word "framework" over and over but none of our projects are frameworks. 

i'm going to again paste my last two replies here:

> please provide specifics on a scenario where soil's jquery cdn has broke something. in the meantime, stop spreading FUD.

> so... you literally have no example of soil's jquery cdn implementation causing any sort of an issue, yet you demand that we remove the feature?

**Comment by danieliser** [comment]
Just to clarify. You chose to make a WP theme. If you didn't like following guidelines then why not just build your own independent products. You guys are obviously hard set on the issue so no point in arguing with a wall.

Just consider your not the only product going to be activated going forward. The guidelines are not there to prevent you from doing what you want. They are there so that everyone can have a great experience without conflict.

@kalenjohnson - Appreciate the link, curious how you handle no conflict mode since CDN jQuery doesn't by default.

**Comment by kalenjohnson** [comment]
> @kalenjohnson - Appreciate the link, curious how you handle no conflict mode since CDN jQuery doesn't by default.

We don't

**Comment by danieliser** [comment]
@kalenjohnson Any particular reason?

**Comment by retlehs** [comment]
until you can provide an example of our jquery cdn implementation causing an actual issue there is nothing to discuss here ¯\_(ツ)_/¯

**Comment by kalenjohnson** [comment]
WP modifies the jQuery file manually to add in no conflict mode. We all understand what no conflict mode is for, and also why it's not necessary. If a theme or plugin author wants to use MooTools or Prototype alongside jQuery then the answer is simple - enqueue WP's version of jQuery.

Trac discussion about noconflict: https://core.trac.wordpress.org/ticket/16494

What I find interesting in that ticket is discussions about educating developers. There is plenty of education on why not touching jQuery enqueueing in WP is around, and we do agree,  you shouldn't enqueue a random jQuery file in your plugin/theme, that won't ever get updated. You don't need to enqueue an outdated jQuery version because it seems to work. What soil does, however, is best practice by web development standards, and that's what Roots tries to do in all it's projects, even if it conflicts with rules set out by WordPress devs.

**Comment by danieliser** [comment]
@kalenjohnson - TBH You hit the nail on the head. That was the purpose of this ticket in general. Education is what is ultimately needed long term, but if a new theme author comes into this "Starter Theme" (@retlehs) and see's you add_theme_support('soil-jquery-cdn'); Most won't self educate themselves on the issue, all they will see is this awesome theme does it, so should I.

Maybe a better alternative is to contribute your loading of jQuery as a core option in the future. That way it CDN loads and auto falls back to local if needed. Win / Win, you guys get what you want, the WP community wins as a whole.

**Comment by danieliser** [comment]
@retlehs, @kalenjohnson - As I said, I didn't start this thread to create a flame war, in fact @kalenjohnson has convinced me to even give it a shot in my own site to test for any real benefit over solid caching & my already existant CDN. That aside I still love what you have created, but you should understand that I deal with dozens of issues related to this issue a week, I can't do anything about it short of helping them fix their themes, so this issue in general has cost me hundreds if not thousands of hours over the last 5 years of supporting my plugins. And I am just one developer in an ocean of plugin & theme devs dealing with these on a daily basis.

So to see a well respected theme doing it just because, kinda burns me a bit. Sure I know how to fix my own site should something go wrong, but we are in the 1% ultimately, the other 99% are completely dependent on us to make things as compatible as we can, so I would implore you to start an issue on WordPress trac to change the core to use your (Super Awesome) CDN Loading with Local Fallback.

In fact if you went that route you may open the doors for other themes to load their own copies with a local fallback as the default whether they set it up or not. Then Everyone wins.

**Comment by QWp6t** [comment]
You could just .... not use the module.

**Comment by kalenjohnson** [comment]
@danieliser we've actually been discussing adding noconflict mode back in when using the soil module. It would bring back parity to the WP version, which in certain scenarios could be helpful.

Basically everyone on the Roots team is rather bashful about making core trac issues. Chances are it'll sit for quite a few years while everyone humms and hawws about it. Or more likely it's ignored and/or closed. Unless something has changed, I don't believe WP has any desire to load jQuery from a CDN.

@QWp6t gj brah

**Comment by danieliser** [comment]
@QWp6t Sure that's a simple decision for a well informed, well established WP dev. But not to someone just getting started. I did as most do and learned everything from blogs and looking directly at how others were doing stuff in their code. I'm willing to bet that is the same for most new WP devs out there.

Seeing a popular theme doing something like this is a sure fire way to get them doing it. Likely they won't dig in to your code deep enough to understand the hows and whys, they simply see CDN jquery and then goto google.

I'm not against people doing it, but it should not be the "norm". It should be enabled specifically by the dev that will ultimately be responsible for supporting it. IE not on by default. Lets get real, most of the people using WP wouldn't know the difference between jQuery and Bootstrap. That said this theme is likely not for them. My concern isn't directly with your theme, as 90% of your users are devs and can fix issues that arise, I'm concerned about the other 10% that will go on to write semi popular themes on the free repo for those millions of other users. Perpetuating bad habits (whether your way is solid or not, if they don't copy your method exactly then what).

All of that said I understand your teams stance, but thats just not the case across the board.

@kalenjohnson I can see where you guys would be coming from, over the last 7 years I have seen many trac tickets come and go without being included.

But over the last 2 years there has been a huge shift in the entire ecosystem of the core teams. Core team growth has been exponential and the commitments this year alone to release a new version every quarter means a lot of tickets are getting pushed through very quickly.

If you guys would consider it I would be happy to lobby on your behalf to the many plugin & theme authors that would approve of it simply to prevent other themes from hacking it themselves. Obviously the technique isn't going anywhere and is becoming more of a fine line to walk. Having core opt for early adoptance of a feature that lets face it has no backward compatibility issues would mean the next generation of themes would avoid this by default.

That is ultimately a viable future solution to the problem.

**Comment by QWp6t** [comment]
> I simply want a valid discussion on the matter

We already had one. You just weren't there for it. But our discussion is what resulted in the module you see here, a module that so far you haven't actually been able to cite a single example of it yielding compatibility or other issues with plugins and/or themes.

> It's pure and simple not in the spirit of WP 

Good.

> I deal with dozens of issues related to this issue a week, I can't do anything about it short of helping them fix their themes, so this issue in general has cost me hundreds if not thousands of hours over the last 5 years of supporting my plugins

And yet you couldn't cite a single instance where Soil's module was the culprit. That's because there's nothing wrong with what we're doing.

> That is ultimately a viable future solution to the problem.

What **problem**? The one you say exists without any actual evidence or data to back up your claim?

The fact here is that you were upset that your issue was closed without discussion. But it was for a valid reason. This isn't worth discussing.You thought we weren't using the same version of jQuery as WP Core, but we are. So there's no problem here. 

As for unwitting users loading our jQuery module (that causes no problems at all but for whatever reason you're still pretending it does), that's a stretch. The user would have to find Soil, which isn't available on wp.org. Then they'd have to install it and add support for it to the theme. If after all of that, it breaks stuff (which there's no evidence to suggest it would), the user is now equipped with the knowledge required to remove support from the theme or simply deactivate the plugin. Like I said, this is a stretch at best. If these bizarre outliers are your idea of a "valid discussion," then that's perfect justification for this issue being closed so abruptly.

But sure, feel free to ask WP Core to add the CDN with local fallback. They should have done that years ago.

**Comment by danieliser** [comment]
@QWp6t - Actually your dead wrong on nearly all accounts. Just because I can't point to a specific site using your theme, doesn't mean that this problem isn't one in general in the WP community. Doesn't go away just cause you say so. Wish that were the case, but then I wouldn't have a job as every plugin & theme would work flawlessly.

Second never said it was specific to Roots or Soil. That was you and your team that made the assumption.

And If I was upset about that fact, I would have simply spammed a link to this thread to dozens of authors who would have jumped in to spam add their 2 cents. Thats not what I did at all. And if you don't think thats doable try the jQuery git issues where the exact same thing happened last week, they disreguarded an issue that pertains to the masses and the masses jumped in and help set the record straight. Now jQuery core team is communicating directly with WP core team to prevent any future issues.

You don't seem to understand the spirit of open source in general, which is quite sad. Once again all of your arguments are based on me finger pointing -> "Bad Roots theme, bad" which never happened.

But as I have mentioned multiple times, and here is a direct quote since you seemed to have missed it:

> I'm not against people doing it, but it should not be the "norm". It should be enabled specifically by the dev that will ultimately be responsible for supporting it. IE not on by default. 

> My concern isn't directly with your theme, as 90% of your users are devs and can fix issues that arise.

I have already digressed and even admitted that your way seems pretty solid, even went as far as to recommend it should be part of core, so @QWp6t, get off your high horse and realize that not everyone is as perfect as you are, that includes 90% of the theme developers out there that consistently release crap, so I'm sorry if you don't see the need simply because you never see the issue and code every line of your sites from scratch, but you and I are not the average user or theme developer are we. But unfortunately for me and most of the WP product dev community this is a real pain point. Shame you are stuck in a bubble and not willing to do anything about your convictions. Seems if you really felt that way about it you would be on wordpress trac and slack trying to get movement like any other dev does to get things done. But you must be to busy with being perfect, won't bother you again.

In any case @kalenjohnson it was nice chatting with you. Have a great day. Close and delete this issue for all I care.
